go/vt/vtgate/planbuilder/select2.go

```diff
@@ -0,0 +1,367 @@
+	switch expr := tableExpr.Expr.(type) {
+	case *sqlparser.TableName:
+		route, table, err := getTablePlan(expr, schema)
+		if err != nil {
+			return nil, nil, err
+		}
+		planBuilder := &RouteBuilder{
+			From:  tableExpr,
+			order: 1,
+			Route: route,
+		}
+		alias := expr.Name
+		if tableExpr.As != "" {
+			alias = tableExpr.As
+		}
+		symbols := NewSymbolTable(alias, table, planBuilder)
+		return planBuilder, symbols, nil
+	case *sqlparser.Subquery:
+		// TODO(sougou): implement.
+		return nil, nil, errors.New("no subqueries")
+	}
+	panic("unreachable")
+}
+
+// getTablePlan produces the initial Route for the specified TableName.
+// It also returns the associated vschema info (*Table) so that
+// it can be used to create the symbol table entry.
+func getTablePlan(tableName *sqlparser.TableName, schema *Schema) (*Route, *Table, error) {
+	if tableName.Qualifier != "" {
+		// TODO(sougou): better error message.
+		return nil, nil, errors.New("tablename qualifier not allowed")
+	}
+	table, reason := schema.FindTable(string(tableName.Name))
+	if reason != "" {
+		return nil, nil, errors.New(reason)
+	}
+	if table.Keyspace.Sharded {
+		return &Route{
+			PlanID:   SelectScatter,
+			Keyspace: table.Keyspace,
+		}, table, nil
+	}
+	return &Route{
+		PlanID:   SelectUnsharded,
+		Keyspace: table.Keyspace,
+	}, table, nil
+}
+
+// processJoin produces a PlanBuilder subtree and SymbolTable
+// for the given Join. If the left and right nodes can be part
+// of the same route, then it's a RouteBuilder. Otherwise,
+// it's a JoinBuilder.
+func processJoin(join *sqlparser.JoinTableExpr, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	switch join.Join {
+	case sqlparser.JoinStr, sqlparser.StraightJoinStr, sqlparser.LeftJoinStr:
+	default:
+		// TODO(sougou): better error message.
+		return nil, nil, errors.New("unsupported join")
+	}
+	lplanBuilder, lsymbols, err := processTableExpr(join.LeftExpr, schema)
+	if err != nil {
+		return nil, nil, err
+	}
+	rplanBuilder, rsymbols, err := processTableExpr(join.RightExpr, schema)
+	if err != nil {
+		return nil, nil, err
+	}
+	switch lplanBuilder := lplanBuilder.(type) {
+	case *JoinBuilder:
+		return makeJoinBuilder(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join)
+	case *RouteBuilder:
+		switch rplanBuilder := rplanBuilder.(type) {
+		case *JoinBuilder:
+			return makeJoinBuilder(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join)
+		case *RouteBuilder:
+			return joinRoutes(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join)
+		}
+	}
+	panic("unreachable")
+}
+
+// makeJoinBuilder creates a new JoinBuilder node out of the two builders.
+// This function is called when the two builders cannot be part of
+// the same route.
+func makeJoinBuilder(lplanBuilder PlanBuilder, lsymbols *SymbolTable, rplanBuilder PlanBuilder, rsymbols *SymbolTable, join *sqlparser.JoinTableExpr) (PlanBuilder, *SymbolTable, error) {
+	err := lsymbols.Add(rsymbols)
+	if err != nil {
+		return nil, nil, err
```

Comment:
As we agreed offline -- add a global comment in the planbuilder file that functions that return errors may change their passed-by-reference parameters and the caller should not rely on them being un-corrupted.